### PR TITLE
8292761: x86: Clone nodes to match complex rules

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch tries to clone a node if it can be matched as a part of a BMI and lea pattern. This may reduce the live range of a local or remove that local completely.

Please take a look and have some reviews. Thanks a lot.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292761](https://bugs.openjdk.org/browse/JDK-8292761): x86: Clone nodes to match complex rules

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9977/head:pull/9977` \
`$ git checkout pull/9977`

Update a local copy of the PR: \
`$ git checkout pull/9977` \
`$ git pull https://git.openjdk.org/jdk pull/9977/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9977`

View PR using the GUI difftool: \
`$ git pr show -t 9977`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9977.diff">https://git.openjdk.org/jdk/pull/9977.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
The patch also shows some improvements regarding benchmarking. The results are as follow:

                                          Before           After
    Benchmark            Mode  Cnt    Score    Error    Score   Error  Units
    CloneNodes.testAndn  avgt   15  374.854 ±  1.439  283.440 ± 3.011  ns/op
    CloneNodes.testLea   avgt   15  448.468 ± 10.071  305.273 ± 1.973  ns/op

The benchmark code is:

    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @State(Scope.Benchmark)
    @Warmup(iterations = 5, time = 1)
    @Measurement(iterations = 5, time = 1)
    @Fork(3)
    public class CloneNodes {
        static final int[] x = new int[1024];
        int y, z;

        @Benchmark
        public void testLea(Blackhole bh) {
            for (int i = 0; i < x.length; i++) {
                int temp = x[i] << 2;
                bh.consume(temp + 3);
                bh.consume(temp + 7);
            }
        }

        @Benchmark
        public void testAndn(Blackhole bh) {
            int y = this.y; int z = this.z;
            for (int i = 0; i < x.length; i++) {
                int temp = ~x[i];
                bh.consume(temp & y);
                bh.consume(temp & z);
            }
        }
    }

I am not sure if this should be included in the patch or it can be included in another existing one. Please let me know your opinions, thanks a lot.

**Comment by TobiHartmann** [comment]
Please include the benchmark in the patch. Could you show the generated code before/after? Thanks!

**Comment by merykitty** [comment]
@TobiHartmann Thanks a lot for your review, I have added the benchmark to the patch, the generated code is as follow:

    CloneNodes::testAndn:
    Baseline:
    movl    $0xffffffff, %r8d
    xorl    0x20(%rbx,%r11,4), %r8d
    movl    %r8d, %r9d
    andl    %edi, %r9d
    andl    %ecx, %r8d

    Patched:
    movl    0x10(%rdi,%r11,4), %r10d
    andnl   %ebx, %r10d, %r9d
    andnl   %ecx, %r10d, %r8d


    CloneNodes::testLea:
    Baseline:
    movl    0x14(%rdi,%r11,4), %r8d
    shll    $0x2, %r8d
    movl    %r8d, %r9d
    addl    $0x7, %r9d
    addl    $0x3, %r8d

    Patched:
    movl    0x10(%rcx,%r11,4), %r10d
    leal    0x3(,%r10,4), %r9d
    leal    0x7(,%r10,4), %r10d

**Review by TobiHartmann — COMMENTED** [review]
Thanks. This looks reasonable to me but please add comments explaining the individual patterns with an example.

Someone more familiar with this code (@jatin-bhateja, @sviswa7 ?), should also have a look.

File: src/hotspot/cpu/x86/x86.ad [hunk 1]
```diff
@@ -2415,26 +2415,101 @@ class FusedPatternMatcher {
...
+      return true;
+    }
+    if (VM_Version::supports_fast_3op_lea() && m->Opcode() == Op_AddI &&
+        check_all_uses(m, [](Node* n){
+          return n->Opcode() == Op_AddI && n->in(2)->Opcode() == Op_ConI;
+        })) {
+      return true;
+    }
+  }
+  if (n->Opcode() == Op_AddL) {
```
Comment (chhagedorn) [inline src/hotspot/cpu/x86/x86.ad:2445]: Could be an `else if`

**Comment by merykitty** [comment]
Thanks @TobiHartmann @chhagedorn for your comments, I have updated the PR to address those.

**Comment by sviswa7** [comment]
@merykitty Could you please add benchmark for lea n, [a + b + i]?

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -2415,26 +2415,113 @@ class FusedPatternMatcher {
...
+    // See above
+    if (n->Opcode() == Op_AndL && m->Opcode() == Op_XorL &&
+        m->in(2)->find_long_con(0) == -1 &&
+        check_all_uses(m, [](Node* n){ return n->Opcode() == Op_AndL; })) {
+      return true;
     }
+    if (m->Opcode() != Op_LoadL) {
+      Node* other = n->in((n->in(1) == m) ? 2 : 1);
+      m = other;
+    }
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:2504]: The caller pd_clone_node is checking if m needs to be cloned, but we are changing m here locally before doing the pattern match? The original code was not doing this. Would it now result in cloning the wrong node?

**Comment by sviswa7** [comment]
@merykitty Do you plan to get this optimization in? I had only one comment (October 18). Please take a look.

**Comment by merykitty** [comment]
@sviswa7 I did some more testing and it seems that this approach does not work reliably in cases of `blsi` and the likes. I am thinking more about this problem. Thanks.

**Pull request closed without merging**